presto-hive/src/main/java/com/facebook/presto/hive/HiveRecordSink.java

```diff
@@ -319,4 +328,36 @@ public static boolean isTypeSupported(Type type)
             return false;
         }
     }
+
+    private Object translateDateTime(Object value)
+    {
+        if (value instanceof SqlDate) {
+            return new Date(TimeUnit.DAYS.toMillis(((SqlDate) value).getDays()));
+        }
+        else if (value instanceof SqlTimestamp) {
+            return new Timestamp(((SqlTimestamp) value).getMillisUtc());
+        }
+
+        return value;
+    }
+
+    private Object translateDateTime(Type type, Object value)
+    {
+        if (isArrayType(type)) {
+            List<Object> newValue = new ArrayList<>();
+            for (Object val : (List<?>) value) {
+                newValue.add(translateDateTime(val));
+            }
+            return newValue;
+        }
+        if (isMapType(type)) {
+            Map<Object, Object> newValue = new HashMap<>();
+            for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
+                newValue.put(translateDateTime(entry.getKey()), translateDateTime(entry.getValue()));
```

Comment:
There could be a second level of recursion required here, since type.getValueType() could be an array, so I think you need to call this recursively with the key and value types